x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,309 @@
+                        phase.getName(), allocateAction);
+                } else {
+                    // remove the action altogether
+                    actionMap.remove(allocateAction.getWriteableName());
+                    logger.debug("ILM policy [{}], phase [{}]: removed the allocate action", lifecyclePolicy.getName(),
+                        phase.getName());
+                }
+
+                Phase updatedPhase = new Phase(phase.getName(), phase.getMinimumAge(), actionMap);
+                Map<String, Phase> updatedPhases =
+                    new HashMap<>(newLifecyclePolicy == null ? lifecyclePolicy.getPhases() : newLifecyclePolicy.getPhases());
+                updatedPhases.put(phaseEntry.getKey(), updatedPhase);
+                newLifecyclePolicy = new LifecyclePolicy(lifecyclePolicy.getName(), updatedPhases);
+            }
+        }
+        return newLifecyclePolicy;
+    }
+
+    static boolean allocateActionDefinesRoutingRules(String nodeAttrName, @Nullable AllocateAction allocateAction) {
+        return allocateAction != null && (allocateAction.getRequire().get(nodeAttrName) != null ||
+            allocateAction.getInclude().get(nodeAttrName) != null ||
+            allocateAction.getExclude().get(nodeAttrName) != null);
+    }
+
+    static List<String> migrateIndices(Metadata.Builder mb, ClusterState currentState, String nodeAttrName) {
+        List<String> migratedIndices = new ArrayList<>();
+        String nodeAttrIndexRequireRoutingSetting = INDEX_ROUTING_REQUIRE_GROUP_SETTING.getKey() + nodeAttrName;
+        String nodeAttrIndexIncludeRoutingSetting = INDEX_ROUTING_INCLUDE_GROUP_SETTING.getKey() + nodeAttrName;
+        for (ObjectObjectCursor<String, IndexMetadata> index : currentState.metadata().indices()) {
+            IndexMetadata indexMetadata = index.value;
+            Settings currentSettings = indexMetadata.getSettings();
+            Settings newSettings = migrateIndexSettings(nodeAttrIndexRequireRoutingSetting, indexMetadata);
+            if (newSettings.equals(currentSettings)) {
+                // migrating based on the `require` setting was not successful so let's check if the index used the `include` routing
+                // setting to configure the allocations and try to migrate it
+                newSettings = migrateIndexSettings(nodeAttrIndexIncludeRoutingSetting, indexMetadata);
+            }
+
+            if (newSettings.equals(currentSettings) == false) {
+                mb.put(IndexMetadata.builder(indexMetadata)
+                    .settings(newSettings)
+                    .settingsVersion(indexMetadata.getSettingsVersion() + 1));
+                migratedIndices.add(indexMetadata.getIndex().getName());
+            }
+        }
+        return migratedIndices;
+    }
+
+    private static Settings migrateIndexSettings(String attributeBasedRoutingSettingName, IndexMetadata indexMetadata) {
```

Comment:
I think it makes sense to add javadocs to these (all the static methods in this class), even if very short, just to help future readers determine what they are doing and what they return